File: C-Plus-Plus/math/Perfect_Number.cpp
/*
Perfect number is a positive integer which is equal to the sum of its proper positive divisors.
For example: 6 is the first perfect number
Proper divisors of 6 are 1, 2, 3
Sum of its proper divisors = 1 + 2 + 3 = 6.
Hence 6 is a perfect number.
*/

#include <bits/stdc++.h>
using namespace std;

int perfectnumber(int n)
{
    int sum=0;
    for (int i = 1; i < n; i++)
    {
        if (n % i == 0)
            sum += i;
    }
    return sum;
}

int main()
{
    int n;
    cout<<"Enter a number to check whether it is a perfect number or not: ";
    cin >> n;
    if (perfectnumber(n) == n)
        cout << n << " is a perfect number." << endl;
    else
        cout << n << " is not a perfect number." << endl;
    return 0;
}

/*

Time Complexity: O(n)
Space Complexity: O(1)

sample 1:
Input: Enter a number to check whether it is a perfect number or not: 6
Output: 6 is a perfect number.

sample 2:
Input: Enter a number to check whether it is a perfect number or not: 7
output: 7 is not a perfect number.

*/ 

File: Dart/maths/PerfectNumber.dart
/*
Perfect number is a number which is equal to the sum of its divisors except the number itself
For example consider 6 
Divisors of 6 are 1, 2, 3
Sum of divisors = 1 + 2 + 3 = 6.
*/

import 'dart:io';

// Check if number is perfect or no
bool isPerfect(int number) {
  int sum = 0;

  for (int index = 1; index < number; index++) {
    // if number is a factor add it to sum
    if (number % index == 0) {
      sum += index;
    }
  }

  if (sum == number) {
    return true;
  }
  return false;
}

// Main Function, with driver code
void main() {
  print("Enter a number:");
  int input = int.parse(stdin.readLineSync()!);

  // Call isPerfect function on input
  if (isPerfect(input)) {
    print("$input is a perfect number");
  } else {
    print("$input is not a perfect number");
  }
}
/**
Time Complexity: O(n)
Space Complexity: O(1)

Sample input/output:
Enter a number:
6
6 is a perfect number

Enter a number:
12
12 is not a perfect number
*/

